core/src/main/java/org/infinispan/reactive/publisher/impl/ClusterPublisherManagerImpl.java

```diff
@@ -936,24 +937,25 @@ private boolean hasWriteBehindSharedStore(PersistenceConfiguration persistenceCo
 
                   int targetBatchSize = (publisher.batchSize + concurrentPublishers - 1) / concurrentPublishers;
 
-                  Publisher<R>[] publisherArray = new Publisher[concurrentPublishers];
+                  InnerPublisherSubscription.InnerPublisherSubscriptionBuilder<K, I, R> builder =
+                        new InnerPublisherSubscription.InnerPublisherSubscriptionBuilder<>(this, targetBatchSize,
+                              targetSupplier, excludedKeys, currentTopology);
+
+                  Publisher<E>[] publisherArray = new Publisher[concurrentPublishers];
                   for (int i = 0; i < concurrentPublishers - 1; ++i) {
-                     publisherArray[i] = InnerPublisherSubscription.createPublisher(this, targetBatchSize, targetSupplier,
-                           excludedKeys, currentTopology);
+                     publisherArray[i] = subToFlowableFunction.apply(builder, null);
                   }
                   // Submit the local target last if necessary (otherwise is a normal submission)
                   // This is done last as we want to send all the remote requests first and only process the local
                   // container concurrently with the remote requests
                   if (localSegments != null) {
-                     publisherArray[concurrentPublishers - 1] = InnerPublisherSubscription.createPublisher(this, targetBatchSize,
-                           targetSupplier, excludedKeys, currentTopology, new AbstractMap.SimpleEntry<>(localAddress, localSegments));
+                     publisherArray[concurrentPublishers - 1] = subToFlowableFunction.apply(builder, new AbstractMap.SimpleEntry<>(localAddress, localSegments));
```

Comment:
> I believe you could avoid the Map.Entry parameter by putting the targets in the list and putting the local segments at the end of the list. It would change the semantics a bit though, because the local segments would be handled last, instead of 4th.

This will increase latency quite a bit for the first entry returned. That is the entire reason for this argument so the current thread can start processing the local entries right away. This in turn provides an entry as soon as possible while still sending remote calls to fill other buffers in parallel.

> I started making the changes locally, but I got distracted as I kept trying to also switch from mergetArray() to flatMap() here, so that InnerPublisherSubscription can process just one target instead of having to switch to a different one.

We don't even have to do that. As the retry mechanism will retry later. If you want we can just calculate the first 4 targets to process in parallel and let it repeat recalculating them for each subsequent run. I personally feel this type of change should be in a different PR though if we want to do that. This PR is about adding the segments to the notifications, the other changes are corollary to the actual change here.